File: server_python/storage.py
from typing import Dict, List, Optional
from datetime import datetime
import uuid
from schemas import Table, TableCreate, Bill, BillCreate, BillItem, BillItemCreate, Payment, PaymentCreate, BillWithItems, DashboardTable

class IStorage:
    """Abstract storage interface"""
    
    # Tables
    async def create_table(self, table: TableCreate) -> Table: ...
    async def get_table(self, id: str) -> Optional[Table]: ...
    async def get_table_by_number(self, number: int, restaurant_name: str) -> Optional[Table]: ...
    async def get_all_tables(self) -> List[Table]: ...
    async def update_table(self, id: str, updates: dict) -> Optional[Table]: ...
    
    # Bills
    async def create_bill(self, bill: BillCreate) -> Bill: ...
    async def get_bill(self, id: str) -> Optional[Bill]: ...
    async def get_bill_by_table_id(self, table_id: str) -> Optional[Bill]: ...
    async def get_bill_with_items(self, id: str) -> Optional[BillWithItems]: ...
    async def update_bill(self, id: str, updates: dict) -> Optional[Bill]: ...
    async def get_all_active_bills(self) -> List[Bill]: ...
    
    # Bill Items
    async def create_bill_item(self, item: BillItemCreate) -> BillItem: ...
    async def get_bill_items(self, bill_id: str) -> List[BillItem]: ...
    async def update_bill_item(self, id: str, updates: dict) -> Optional[BillItem]: ...
    async def get_bill_item(self, id: str) -> Optional[BillItem]: ...
    
    # Payments
    async def create_payment(self, payment: PaymentCreate) -> Payment: ...
    async def get_payments_by_bill_id(self, bill_id: str) -> List[Payment]: ...
    
    # Dashboard
    async def get_dashboard_tables(self) -> List[DashboardTable]: ...

class MemStorage(IStorage):
    """In-memory storage implementation"""
    
    def __init__(self):
        self.tables: Dict[str, Table] = {}
        self.bills: Dict[str, Bill] = {}
        self.bill_items: Dict[str, BillItem] = {}
        self.payments: Dict[str, Payment] = {}
        self._initialize_sample_data()
    
    def _initialize_sample_data(self):
        """Initialize with sample data"""
        # Create sample table
        table_id = str(uuid.uuid4())
        table = Table(
            id=table_id,
            number=7,
            restaurant_name="bella-vista",
            qr_code="https://splitbill.app/t/7/bella-vista",
            is_active=True,
            created_at=datetime.now()
        )
        self.tables[table_id] = table
        
        # Create sample bill
        bill_id = str(uuid.uuid4())
        bill = Bill(
            id=bill_id,
            table_id=table_id,
            total="89.50",
            paid="32.75",
            remaining="56.75",
            status="partial",
            guest_count=4,
            start_time=datetime.now(),
            is_active=True
        )
        self.bills[bill_id] = bill
        
        # Create sample bill items
        items_data = [
            {"name": "Caesar Salad", "price": "18.50", "quantity": "1", "paid_quantity": "0"},
            {"name": "Grilled Salmon", "price": "32.00", "quantity": "1", "paid_quantity": "0"},
            {"name": "Wine Bottle (Shared)", "price": "45.00", "quantity": "1", "paid_quantity": "0.5"},
            {"name": "Dessert", "price": "12.00", "quantity": "1", "paid_quantity": "0"},
        ]
        
        for item_data in items_data:
            item_id = str(uuid.uuid4())
            item = BillItem(
                id=item_id,
                bill_id=bill_id,
                name=item_data["name"],
                price=item_data["price"],
                quantity=item_data["quantity"],
                paid_quantity=item_data["paid_quantity"]
            )
            self.bill_items[item_id] = item
        
        # Add more sample tables for dashboard
        for i in range(3, 13):
            if i == 7:  # Skip table 7 as it's already created
                continue
            
            t_id = str(uuid.uuid4())
            sample_table = Table(
                id=t_id,
                number=i,
                restaurant_name="bella-vista",
                qr_code=f"https://splitbill.app/t/{i}/bella-vista",
                is_active=True,
                created_at=datetime.now()
            )
            self.tables[t_id] = sample_table
            
            # Create bills for some tables
            if i in [3, 5, 12]:
                b_id = str(uuid.uuid4())
                status = "unpaid"
                paid = "0"
                remaining = "65.25"
                total = "65.25"
                
                if i == 3:
                    status = "paid"
                    paid = "65.25"
                    remaining = "0"
                elif i == 5:
                    status = "partial"
                    paid = "45.60"
                    remaining = "32.80"
                    total = "78.40"
                elif i == 12:
                    status = "unpaid"
                    paid = "0"
                    remaining = "156.75"
                    total = "156.75"
                
                sample_bill = Bill(
                    id=b_id,
                    table_id=t_id,
                    total=total,
                    paid=paid,
                    remaining=remaining,
                    status=status,
                    guest_count=2 if i == 3 else 3 if i == 5 else 6,
                    start_time=datetime.now(),
                    is_active=True
                )
                self.bills[b_id] = sample_bill
    
    async def create_table(self, table: TableCreate) -> Table:
        table_id = str(uuid.uuid4())
        table = Table(
            id=table_id,
            created_at=datetime.now(),
            **table.model_dump()
        )
        self.tables[table_id] = table
        return table
    
    async def get_table(self, id: str) -> Optional[Table]:
        return self.tables.get(id)
    
    async def get_table_by_number(self, number: int, restaurant_name: str) -> Optional[Table]:
        for table in self.tables.values():
            if table.number == number and table.restaurant_name == restaurant_name:
                return table
        return None
    
    async def get_all_tables(self) -> List[Table]:
        return list(self.tables.values())
    
    async def update_table(self, id: str, updates: dict) -> Optional[Table]:
        table = self.tables.get(id)
        if not table:
            return None
        
        for key, value in updates.items():
            if hasattr(table, key):
                setattr(table, key, value)
        
        self.tables[id] = table
        return table
    
    async def create_bill(self, bill: BillCreate) -> Bill:
        bill_id = str(uuid.uuid4())
        bill = Bill(
            id=bill_id,
            start_time=datetime.now(),
            **bill.model_dump()
        )
        self.bills[bill_id] = bill
        return bill
    
    async def get_bill(self, id: str) -> Optional[Bill]:
        return self.bills.get(id)
    
    async def get_bill_by_table_id(self, table_id: str) -> Optional[Bill]:
        for bill in self.bills.values():
            if bill.table_id == table_id and bill.is_active:
                return bill
        return None
    
    async def get_bill_with_items(self, id: str) -> Optional[BillWithItems]:
        bill = self.bills.get(id)
        if not bill:
            return None
        
        items = [item for item in self.bill_items.values() if item.bill_id == id]
        table = self.tables.get(bill.table_id)
        if not table:
            return None
        
        return BillWithItems(
            **bill.model_dump(),
            items=items,
            table=table
        )
    
    async def update_bill(self, id: str, updates: dict) -> Optional[Bill]:
        bill = self.bills.get(id)
        if not bill:
            return None
        
        for key, value in updates.items():
            if hasattr(bill, key):
                setattr(bill, key, value)
        
        self.bills[id] = bill
        return bill
    
    async def get_all_active_bills(self) -> List[Bill]:
        return [bill for bill in self.bills.values() if bill.is_active]
    
    async def create_bill_item(self, item: BillItemCreate) -> BillItem:
        item_id = str(uuid.uuid4())
        item = BillItem(
            id=item_id,
            **item.model_dump()
        )
        self.bill_items[item_id] = item
        return item
    
    async def get_bill_items(self, bill_id: str) -> List[BillItem]:
        return [item for item in self.bill_items.values() if item.bill_id == bill_id]
    
    async def update_bill_item(self, id: str, updates: dict) -> Optional[BillItem]:
        item = self.bill_items.get(id)
        if not item:
            return None
        
        for key, value in updates.items():
            if hasattr(item, key):
                setattr(item, key, value)
        
        self.bill_items[id] = item
        return item
    
    async def get_bill_item(self, id: str) -> Optional[BillItem]:
        return self.bill_items.get(id)
    
    async def create_payment(self, payment: PaymentCreate) -> Payment:
        payment_id = str(uuid.uuid4())
        payment = Payment(
            id=payment_id,
            processed_at=datetime.now(),
            **payment.model_dump()
        )
        self.payments[payment_id] = payment
        return payment
    
    async def get_payments_by_bill_id(self, bill_id: str) -> List[Payment]:
        return [payment for payment in self.payments.values() if payment.bill_id == bill_id]
    
    async def get_dashboard_tables(self) -> List[DashboardTable]:
        dashboard_tables = []
        
        for table in self.tables.values():
            bill = None
            for b in self.bills.values():
                if b.table_id == table.id and b.is_active:
                    bill = b
                    break
            
            items = []
            if bill:
                items = [item for item in self.bill_items.values() if item.bill_id == bill.id]
            
            dashboard_tables.append(DashboardTable(
                id=table.id,
                number=table.number,
                restaurant_name=table.restaurant_name,
                bill=bill,
                items=items,
                guest_count=bill.guest_count or 0 if bill else 0,
                start_time=bill.start_time if bill else None
            ))
        
        # Sort by table number
        dashboard_tables.sort(key=lambda x: x.number)
        return dashboard_tables

# Global storage instance
storage = MemStorage()

File: server/storage.ts
import { type Table, type InsertTable, type Bill, type InsertBill, type BillItem, type InsertBillItem, type Payment, type InsertPayment, type BillWithItems, type DashboardTable } from "@shared/schema";
import { randomUUID } from "crypto";

export interface IStorage {
  // Tables
  createTable(table: InsertTable): Promise<Table>;
  getTable(id: string): Promise<Table | undefined>;
  getTableByNumber(number: number, restaurantName: string): Promise<Table | undefined>;
  getAllTables(): Promise<Table[]>;
  updateTable(id: string, updates: Partial<Table>): Promise<Table | undefined>;

  // Bills
  createBill(bill: InsertBill): Promise<Bill>;
  getBill(id: string): Promise<Bill | undefined>;
  getBillByTableId(tableId: string): Promise<Bill | undefined>;
  getBillWithItems(id: string): Promise<BillWithItems | undefined>;
  updateBill(id: string, updates: Partial<Bill>): Promise<Bill | undefined>;
  getAllActiveBills(): Promise<Bill[]>;

  // Bill Items
  createBillItem(item: InsertBillItem): Promise<BillItem>;
  getBillItems(billId: string): Promise<BillItem[]>;
  updateBillItem(id: string, updates: Partial<BillItem>): Promise<BillItem | undefined>;
  getBillItem(id: string): Promise<BillItem | undefined>;

  // Payments
  createPayment(payment: InsertPayment): Promise<Payment>;
  getPaymentsByBillId(billId: string): Promise<Payment[]>;

  // Dashboard
  getDashboardTables(): Promise<DashboardTable[]>;
}

export class MemStorage implements IStorage {
  private tables: Map<string, Table>;
  private bills: Map<string, Bill>;
  private billItems: Map<string, BillItem>;
  private payments: Map<string, Payment>;

  constructor() {
    this.tables = new Map();
    this.bills = new Map();
    this.billItems = new Map();
    this.payments = new Map();
    this.initializeSampleData();
  }

  private initializeSampleData() {
    // Create sample table
    const tableId = randomUUID();
    const table: Table = {
      id: tableId,
      number: 7,
      restaurantName: "Bella Vista Restaurant",
      qrCode: "https://splitbill.app/t/7/bella-vista",
      isActive: true,
      createdAt: new Date(),
    };
    this.tables.set(tableId, table);

    // Create sample bill
    const billId = randomUUID();
    const bill: Bill = {
      id: billId,
      tableId: tableId,
      total: "89.50",
      paid: "32.75",
      remaining: "56.75",
      status: "partial",
      guestCount: 4,
      startTime: new Date(),
      isActive: true,
    };
    this.bills.set(billId, bill);

    // Create sample bill items
    const items = [
      { name: "Caesar Salad", price: "18.50", quantity: "1", paidQuantity: "0" },
      { name: "Grilled Salmon", price: "32.00", quantity: "1", paidQuantity: "0" },
      { name: "Wine Bottle (Shared)", price: "45.00", quantity: "1", paidQuantity: "0.5" },
      { name: "Dessert", price: "12.00", quantity: "1", paidQuantity: "0" },
    ];

    items.forEach(item => {
      const itemId = randomUUID();
      const billItem: BillItem = {
        id: itemId,
        billId: billId,
        name: item.name,
        price: item.price,
        quantity: item.quantity,
        paidQuantity: item.paidQuantity,
      };
      this.billItems.set(itemId, billItem);
    });

    // Add more sample tables for dashboard
    for (let i = 3; i <= 12; i++) {
      if (i === 7) continue; // Skip table 7 as it's already created
      
      const tId = randomUUID();
      const sampleTable: Table = {
        id: tId,
        number: i,
        restaurantName: "Bella Vista Restaurant",
        qrCode: `https://splitbill.app/t/${i}/bella-vista`,
        isActive: true,
        createdAt: new Date(),
      };
      this.tables.set(tId, sampleTable);

      // Create bills for some tables
      if (i === 3 || i === 5 || i === 12) {
        const bId = randomUUID();
        let status = "unpaid";
        let paid = "0";
        let remaining = "65.25";
        let total = "65.25";

        if (i === 3) {
          status = "paid";
          paid = "65.25";
          remaining = "0";
        } else if (i === 5) {
          status = "partial";
          paid = "45.60";
          remaining = "32.80";
          total = "78.40";
        } else if (i === 12) {
          status = "unpaid";
          paid = "0";
          remaining = "156.75";
          total = "156.75";
        }

        const sampleBill: Bill = {
          id: bId,
          tableId: tId,
          total,
          paid,
          remaining,
          status,
          guestCount: i === 3 ? 2 : i === 5 ? 3 : 6,
          startTime: new Date(Date.now() - Math.random() * 3600000), // Random time within last hour
          isActive: true,
        };
        this.bills.set(bId, sampleBill);
      }
    }
  }

  async createTable(insertTable: InsertTable): Promise<Table> {
    const id = randomUUID();
    const table: Table = {
      ...insertTable,
      id,
      createdAt: new Date(),
    };
    this.tables.set(id, table);
    return table;
  }

  async getTable(id: string): Promise<Table | undefined> {
    return this.tables.get(id);
  }

  async getTableByNumber(number: number, restaurantName: string): Promise<Table | undefined> {
    return Array.from(this.tables.values()).find(
      table => table.number === number && table.restaurantName === restaurantName
    );
  }

  async getAllTables(): Promise<Table[]> {
    return Array.from(this.tables.values());
  }

  async updateTable(id: string, updates: Partial<Table>): Promise<Table | undefined> {
    const table = this.tables.get(id);
    if (!table) return undefined;
    
    const updatedTable = { ...table, ...updates };
    this.tables.set(id, updatedTable);
    return updatedTable;
  }

  async createBill(insertBill: InsertBill): Promise<Bill> {
    const id = randomUUID();
    const bill: Bill = {
      ...insertBill,
      id,
      startTime: new Date(),
    };
    this.bills.set(id, bill);
    return bill;
  }

  async getBill(id: string): Promise<Bill | undefined> {
    return this.bills.get(id);
  }

  async getBillByTableId(tableId: string): Promise<Bill | undefined> {
    return Array.from(this.bills.values()).find(bill => bill.tableId === tableId && bill.isActive);
  }

  async getBillWithItems(id: string): Promise<BillWithItems | undefined> {
    const bill = this.bills.get(id);
    if (!bill) return undefined;

    const items = Array.from(this.billItems.values()).filter(item => item.billId === id);
    const table = this.tables.get(bill.tableId);
    if (!table) return undefined;

    return { ...bill, items, table };
  }

  async updateBill(id: string, updates: Partial<Bill>): Promise<Bill | undefined> {
    const bill = this.bills.get(id);
    if (!bill) return undefined;
    
    const updatedBill = { ...bill, ...updates };
    this.bills.set(id, updatedBill);
    return updatedBill;
  }

  async getAllActiveBills(): Promise<Bill[]> {
    return Array.from(this.bills.values()).filter(bill => bill.isActive);
  }

  async createBillItem(insertItem: InsertBillItem): Promise<BillItem> {
    const id = randomUUID();
    const item: BillItem = {
      ...insertItem,
      id,
    };
    this.billItems.set(id, item);
    return item;
  }

  async getBillItems(billId: string): Promise<BillItem[]> {
    return Array.from(this.billItems.values()).filter(item => item.billId === billId);
  }

  async updateBillItem(id: string, updates: Partial<BillItem>): Promise<BillItem | undefined> {
    const item = this.billItems.get(id);
    if (!item) return undefined;
    
    const updatedItem = { ...item, ...updates };
    this.billItems.set(id, updatedItem);
    return updatedItem;
  }

  async getBillItem(id: string): Promise<BillItem | undefined> {
    return this.billItems.get(id);
  }

  async createPayment(insertPayment: InsertPayment): Promise<Payment> {
    const id = randomUUID();
    const payment: Payment = {
      ...insertPayment,
      id,
      processedAt: new Date(),
    };
    this.payments.set(id, payment);
    return payment;
  }

  async getPaymentsByBillId(billId: string): Promise<Payment[]> {
    return Array.from(this.payments.values()).filter(payment => payment.billId === billId);
  }

  async getDashboardTables(): Promise<DashboardTable[]> {
    const dashboardTables: DashboardTable[] = [];
    
    for (const table of this.tables.values()) {
      const bill = Array.from(this.bills.values()).find(b => b.tableId === table.id && b.isActive);
      const items = bill ? Array.from(this.billItems.values()).filter(item => item.billId === bill.id) : [];
      
      dashboardTables.push({
        id: table.id,
        number: table.number,
        restaurantName: table.restaurantName,
        bill: bill || null,
        items,
        guestCount: bill?.guestCount || 0,
        startTime: bill?.startTime || null,
      });
    }
    
    return dashboardTables.sort((a, b) => a.number - b.number);
  }
}

export const storage = new MemStorage();
